source/hp/dof_handler.cc

```diff
@@ -1522,22 +1520,78 @@ namespace hp
     tria_listeners.push_back(this->tria->signals.create.connect(std::bind(
       &DoFHandler<dim, spacedim>::post_refinement_action, std::ref(*this))));
 
-    // Only connect the update of active fe indices if we are not using a
-    // p::d::Triangulation. In this case, the class ActiveFEIndicesTransfer
-    // has to be consulted.
-    if (dynamic_cast<const parallel::distributed::Triangulation<dim, spacedim>
-                       *>(&*this->tria) == nullptr)
+    // decide whether we need a sequential or a parallel shared/distributed
+    // policy and attach corresponding callback functions dealing with the
+    // transfer of active_fe_indices
+    if (const auto *distributed_tria = dynamic_cast<
+          const parallel::distributed::Triangulation<dim, spacedim> *>(
+          &this->get_triangulation()))
       {
-        tria_listeners.push_back(this->tria->signals.pre_refinement.connect(
-          std::bind(&DoFHandler<dim, spacedim>::pre_refinement_fe_index_update,
-                    std::ref(*this))));
-        tria_listeners.push_back(this->tria->signals.post_refinement.connect(
-          std::bind(&DoFHandler<dim, spacedim>::post_refinement_fe_index_update,
-                    std::ref(*this))));
+        policy = std_cxx14::make_unique<
+          internal::DoFHandlerImplementation::Policy::ParallelDistributed<
+            DoFHandler<dim, spacedim>>>(*this);
+
+#ifdef DEAL_II_WITH_P4EST
+        cell_data_transfer = std_cxx14::make_unique<
```

Comment:
why is this necessary? does it not compile if you don't have P4EST configured?

Reply:
The implementation of the `CellDataTransfer` class is currently embedded in a corresponding `DEAL_II_WITH_P4EST` preprocessor conditional, including its constructor.

The linker complains when `deal.II` is not configured with `p4est`.